Content.Server/MachineLinking/System/SignalLinkerSystem.cs

```diff
@@ -20,262 +20,192 @@ namespace Content.Server.MachineLinking.System
+                    else if (!receiverPort.Transmitters.Contains(transmitterPort))
+                        receiverPort.Transmitters.Add(transmitterPort);
         }
 
         private void OnTransmitterRemoved(EntityUid uid, SignalTransmitterComponent component, ComponentRemove args)
         {
-            _linkCollection.RemoveLinks(component);
+            foreach (var transmitterPort in component.Outputs)
+                foreach (var (receiver, receiverPortName) in transmitterPort.Receivers)
+                    if (TryComp(receiver, out SignalReceiverComponent? receiverComponent) &&
+                        receiverComponent.Inputs.TryGetPort(receiverPortName, out var receiverPort))
+                        receiverPort.Transmitters.Remove(transmitterPort);
         }
 
         private void OnReceiverRemoved(EntityUid uid, SignalReceiverComponent component, ComponentRemove args)
         {
-            _linkCollection.RemoveLinks(component);
+            foreach (var receiverPort in component.Inputs)
+                foreach (var transmitterPort in receiverPort.Transmitters)
+                    transmitterPort.Receivers.Remove((component.Owner, receiverPort.Name));
         }
 
-        private void OnTransmitterInvokePort(EntityUid uid, SignalTransmitterComponent component, InvokePortEvent args)
+        private void OnTransmitterInteractUsing(EntityUid uid, SignalTransmitterComponent component, InteractUsingEvent args)
         {
-            if (!component.Outputs.TryGetPort(args.Port, out var port)) throw new PortNotFoundException();
-
-            if (args.Value == null)
-            {
-                if (port.Type != null && !port.Type.IsNullable()) throw new InvalidPortValueException();
-            }
-            else
-            {
-                if (port.Type == null || !args.Value.GetType().IsAssignableTo(port.Type))
-                    throw new InvalidPortValueException();
-            }
+            if (args.Handled) return;
 
-            port.Signal = args.Value;
+            if (!TryComp(args.Used, out SignalLinkerComponent? linker) ||
+                !TryComp(args.User, out ActorComponent? actor))
+                return;
 
-            foreach (var link in _linkCollection.GetLinks(component, port.Name))
+            if (component.Outputs.Count == 1)
             {
-                if (!IsInRange(component, link.ReceiverComponent)) continue;
-
-                RaiseLocalEvent(link.ReceiverComponent.Owner,
-                    new SignalReceivedEvent(link.Receiverport.Name, args.Value), false);
+                var port = component.Outputs.First();
+                LinkerTransmitterInteraction(args.User, linker, component, port.Name);
+                args.Handled = true;
+                return;
             }
+
+            var bui = component.Owner.GetUIOrNull(SignalTransmitterUiKey.Key);
+            if (bui == null) return;
```

Comment:
instead of component.Owner, you can just use `uid`.

Also, instead of using the entityUid helper, which is just obscuring a 
```
IoCManager.Resolve<IEntityManager>().GetComponentOrNull<ServerUserInterfaceComponent>(entity)?.GetBoundUserInterfaceOrNull(uiKey);
```
you should use `UserInterfaceSystem.GetUiOrNull()`